### ISSUE TITLE
~zero handled improperly

### ISSUE THREAD

**Issue by jpziegler** [GaloisInc/cryptol#423]
Observe the following commands:
```
Cryptol> let add = \(a:[32],b) -> a + b
Cryptol> add ~zero
0xffffffff
Cryptol> add (~zero)
0xfffffffe
```
The two calls to `add` result in different values, but they should both be `0xfffffffe`.

**Comment by brianhuffman** [comment]
Ha! I know exactly what's going on here. It stems from this declaration in Cryptol.cry:
```
(~) : {a} a -> a
(~) = complement
```
There is already a special case in the parser for unary prefix `~`, but this does something different: It defines `~` as a *binary* *infix* operator. So when you write `add ~zero`, it gets parsed as `(~) add zero`, which means `complement add zero`.

The `complement` is being used here at the higher-order type `(([32],[32]) -> [32]) -> ([32],[32]) -> [32]`. Complementing a function is done pointwise, so `complement add zero` is equivalent to `complement (add zero)`. This explains why you're getting `0xffffffff` as a result.

We should probably just remove that declaration from Cryptol.cry. On the other hand, it's fun to be able to write `~f~ x` in place of `f x`, if you want to emphasize the `f` ;-)

**Comment by jpziegler** [comment]
Ha!

Thank you for the explanation. I agree there should be a change, though, as the current behavior is unexpected and potentially dangerous.

**Comment by brianhuffman** [comment]
Removing the definition of the infix `~` from the Cryptol prelude breaks the regression test for #296, which is about supporting `:t (~)` at the REPL. (Issue #296 was the reason why the infix `~` definition was added in the first place.)

I think we should revert c620cbf, which was the "fix" for #296. `:t (~)` should *not* be supported at the REPL, as the `(op)` syntax is for infix operators, and `~` is not an infix operator.

If users don't know what `~` is syntactic sugar for, they can still find its type with `:t \x -> ~x`. I should also point out that what unary prefix `~` and `-` stand for is easily discoverable with `:t`:
```
Cryptol> :t \x -> ~x
(\x -> complement x) : {a} a -> a
Cryptol> :t ~True
complement True : Bit
Cryptol> :t \x -> -x
(\x -> negate x) : {a} (Arith a) => a -> a
```
